src/libraries/System.Private.CoreLib/src/System/DateTime.Windows.cs

```diff
@@ -1,30 +1,164 @@
+                ((delegate* unmanaged[Stdcall]<ulong*, void>)pfnGetSystemTime)(&filetimeStd);
+                ((delegate* unmanaged[Stdcall]<ulong*, void>)pfnGetSystemTimePrecise)(&filetimePrecise);
+
+                if (Math.Abs((long)(filetimeStd - filetimePrecise)) <= 100 * TicksPerMillisecond)
+                {
+                    pfnGetSystemTime = pfnGetSystemTimePrecise; // use the precise version
+                }
+            }
+
+            return (delegate* unmanaged[Stdcall]<ulong*, void>)pfnGetSystemTime;
+        }
 
         public static unsafe DateTime UtcNow
         {
             get
             {
-                if (s_systemSupportsLeapSeconds)
+                // The OS tick count and .NET's tick count are slightly different. The OS tick
+                // count is the *absolute* number of 100-ns intervals which have elapsed since
+                // January 1, 1601 (UTC). Due to leap second handling, the number of ticks per
+                // day is variable. Dec. 30, 2016 had 864,000,000,000 ticks (a standard 24-hour
+                // day), but Dec. 31, 2016 had 864,010,000,000 ticks due to leap second insertion.
+                // In .NET, *every* day is assumed to have exactly 24 hours (864,000,000,000 ticks).
+                // This means that per the OS, midnight Dec. 31, 2016 + 864 bn ticks = Dec. 31, 2016 23:59:60,
+                // but per .NET, midnight Dec. 31, 2016 + 864 bn ticks = Jan. 1, 2017 00:00:00.
+                //
+                // We can query the OS and have it deconstruct the tick count into (yyyy-mm-dd hh:mm:ss),
+                // constructing a new DateTime object from these components, but this is slow.
+                // So instead we'll rely on the fact that leap seconds only ever adjust the day
+                // by +1 or -1 second, and only at the very end of the day. That is, time rolls
+                // 23:59:58 -> 00:00:00 (negative leap second) or 23:59:59 -> 23:59:60 (positive leap
+                // second). Thus we assume that each day has at least 23 hr 59 min 59 sec, or
+                // 863,990,000,000 ticks.
+                //
+                // We take advantage of this by caching what the OS believes the tick count is at
+                // the beginning of the month vs. what .NET believes the tick count is at the beginning
+                // of the month. When the OS returns a tick count to us, if it's before 23:59:59 on the
+                // last day of the month, then we know that there's no way for a leap second to have been
+                // inserted or removed, and we can short-circuit the leap second handling logic by
+                // performing a quick addition and returning immediately.
+                //
+                // This relies on:
+                // a) Leap seconds only ever taking place on the last day of the month
+                //    (historically these are Jun. & Dec., but we support any month);
```

Comment:
Maybe adding the link https://datacenter.iers.org/data/latestVersion/16_BULLETIN_C16.txt can be a reference to that.